docs/source/ProgrammingGuide/Kokkos-and-Virtual-Functions.md

```diff
@@ -82,87 +91,125 @@ This code is extremely ugly, but leads to functional virtual function calls on t
+This would crash if `someHostValue` is not accessible on the device (e.g. for a small array). The most productive solution we've found in these cases is to allocate the instance in `SharedSpace`, initialize it on the device, and then fill in fields on the host. To wit:
 
 ```c++
-Base* deviceInstance = Kokkos::kokkos_malloc<Kokkos::SharedSpace>(sizeof(Derived));
-Kokkos::parallel_for(... {
-  new((Derived*)deviceInstance) Derived(); // construct an instance in the place the the pointer deviceInstance points to
+auto deviceInstanceMemory = Kokkos::kokkos_malloc<Kokkos::SharedSpace>(sizeof(Derived)); // allocate on shared space
+Kokkos::parallel_for("Initialize", 1, KOKKOS_LAMBDA (const int i) {
+  new (static_cast<Derived*>(deviceInstanceMemory)) Derived();
 });
-deviceInstance->setAField(someHostValue); // set some field on the host
+Kokkos::fence("Wait for initialize");
+Base* deviceInstance = static_cast<Derived*>(deviceInstanceMemory);
+deviceInstance->setAField(someHostValue); // set on host
 ```
 
 This is the solution that the code teams we have talked to have said is the most productive way to solve the problem. Nevertheless, it should be kept in mind, that this restricts virtual function calls to the device.
+In the case presented above, `setAField()` is not a virtual function.
 
 ## But what if I do not really need the V-Tables on the device side?
-Consider the following example which calls the `virtual Bar()` on the device from a pointer of derived class type.
-One might think this should work because no V-Table lookup on the device is neccessary.
+
+Consider the following example which calls the virtual function `Bar()` on the device from a pointer of derived class type.
+One might think this should work because no V-Table lookup on the device is necessary.
+
 ```c++
 #include <Kokkos_Core.hpp>
-#include <cstdio>
 
 struct Base
 {
-    KOKKOS_DEFAULTED_FUNCTION
-    virtual ~Base() = default;
+  KOKKOS_DEFAULTED_FUNCTION
+  virtual ~Base() = default;
 
-    KOKKOS_FUNCTION
-    virtual void Bar() const = 0;
+  KOKKOS_FUNCTION
+  virtual void Bar() const = 0;
 };
 
 struct Derived : public Base
 {
-    KOKKOS_FUNCTION
-    void Bar() const override
-    { printf("Hello from Derived\n"); }
-
-    void apply(){
-        Kokkos::parallel_for("myLoop",10,
-            KOKKOS_CLASS_LAMBDA (const size_t i) { this->Bar(); }
-        );
-    }
+  KOKKOS_FUNCTION
+  void Bar() const override
+  { Kokkos::printf("Hello from Derived\n"); }
+
+  void apply(){
+    Kokkos::parallel_for("myLoop", 10,
+      KOKKOS_CLASS_LAMBDA (const size_t i) { this->Bar(); }
+    );
+  }
 };
 
 int main (int argc, char *argv[])
 {
-    Kokkos::initialize(argc,argv);
-    {
-      auto derivedPtr = std::make_shared<Derived>();
-      derivedPtr->apply();
-      Kokkos::fence();
-    }
-    Kokkos::finalize();
+  Kokkos::ScopeGuard kokkos(argc, argv);
+  auto derivedPtr = std::make_shared<Derived>();
+  derivedPtr->apply();
+  Kokkos::fence();
```

Comment:
why use the `ScopeGuard` here?

Reply:
`ScopeGuard` should be preferred over `initialize`/`finalize` as it ensures that `finalize` is called no matter what (RAII).

Reply:
hmm AFAIK we try to avoid `ScopeGuard` in the examples, both in the tutorials as well as in the documentation. I do agree, that it would not make much of a difference in this example, but I would favor consistency

Reply:
I think it's a good opportunity to start using it more.